test/test_autograd.py

```diff
@@ -4558,6 +4560,101 @@ def parent_on_cpu(inp):
         # gpu thread ReadyQueue
         out.sum().backward()
 
+class TestMultithreadAutograd(TestCase):
+    def _run_py_multithread_fn(self, fn, args=(), num_threads=10, kwargs=None):
+        threads = []
+        for _ in range(num_threads):
+            p = threading.Thread(target=fn, args=(args))
+            p.start()
+            threads.append(p)
+
+        for p in threads:
+            p.join()
+
+
+    def test_simple_backward(self):
+        # simple multithreaded backward that create threads in the beginning of training
+        # and everything else is training separately, i.e. inputs, operations, etc.
+        def train_fn():
+            x = torch.ones(5, 5, requires_grad=True)
+            y = (x + 3) * (x + 4) * 0.5
+            y.sum().backward()
+            self.assertEqual(x.grad, x + 3.5)
+
+        self._run_py_multithread_fn(train_fn)
+
+
+    def test_simple_backward_same_input(self):
+        # simple multithreaded backward with only shared inputs (i.e. Hogwil training with CPU threads)
+        def train_fn_backward(x):
+            y = (x + 3) * (x + 4) * 0.5
+            y.sum().backward()
+
+        x = torch.ones(5, 5, requires_grad=True)
+        self._run_py_multithread_fn(train_fn_backward, (x,))
+        # Since we are calling backward from multiple threads
+        # this will race to .grad since input variable is shared
+        # across thread and it's expected gradients are not correct
+        self.assertNotEqual(x.grad, x + 3.5)
+
+        def train_fn_grad(x):
+            y = (x + 3) * (x + 4) * 0.5
+            grads = torch.autograd.grad(y.sum(), x)
+            self.assertEqual(len(grads), 1)
+            self.assertEqual(grads[0], x + 3.5)
+
+        # since we use functional grad() api, gradients should work as expected
+        self._run_py_multithread_fn(train_fn_grad, (x,))
+
+
+    def test_fork_in_middle(self):
+        # User might write a network that starts on one CPU thread, then runs its second half
+        # concurrently with other threads (either via python threading or fork/join calls),
+        # then calls backward()/grad() on BOTH threads, like a Y pattern from input at the
+        # bottom to output at the top. This way part of the GraphTask is being shared across
+        # different threads and we need to ensure user specify retain_graph=True, otherwise
+        # error out with the correct error message
+
+        # multiple backward with python threads, error with no retain_graph after the first thread
+        def train_fn_no_retain_graph(x):
+            y = x + x * 2
+            try:
+                y.sum().backward()
+            except RuntimeError as error:
+                self.assertRegex(str(error), "Specify retain_graph=True")
```

Comment:
Could you add a variable in the parent scope like `counter = [0, 0]` and then count how many failed and how many worked? Just to be sure that you get the behavior you expect.
Because here, maybe none of them raise the error, or all of them, and we have no way to know.